### DISCUSSION TITLE
Preferred engine per instance, where to store setting

### DISCUSSION THREAD

**Discussion by afbjorklund** [lima-vm/lima#993] [General]
Should we add a value to hold the default container engine, for each instance ?

e.g.
`default`: "containerd" (nerdctl)
`docker`: "docker"
`podman`: "podman"
`ubuntu`: "containerd"
`alpine`: "" (none)
`fedora`: "containerd"
etc

Or what would be the best way to "know" which client to use, for each instance ?

nerdctl.lima
docker.lima
podman.lima
<error>

----

The main use was for the Lima GUI, where it should be possible to interact with it.

* https://github.com/afbjorklund/lima-gui/issues/18

Examples:
* Version
* Info
* Containers
* Images

But also for the CLI, it could be useful with a single wrapper for all three engines.

Basically: `limactl shell $INSTANCE $ENGINE`

**Comment by jandubois** [comment]
I don't understand how it would be used. Is this a data that `limactl ls --format` would expose? In that case I think we need to re-open the `limactl inspect` discussion. 😄 

Anyways, if that is the idea, then I think it would be better to have a more generic mechanism, like

```yaml
metadata:
  engine: containerd
  cli: nerdctl.lima
  ...
```

So you can define whatever additional fields you want. Of course you still need some kind of convention, so that more than one image exposes the same metadata field; otherwise it becomes pretty useless.

**Reply by afbjorklund** [reply]
I'm not sure the best way to get to it, but I guess so - reading the JSON but anyway...

I _could_ input it with a drop-down in the GUI, or "try everything" like is done with CRI.
But I think the best would be if the instance store already contained the information ?

Currently I have just added data sources like `uname`, `uptime` and `/etc/os-release`.

---

Currently these are used: `limactl list --json`

`nerdctl version --format=json`
`nerdctl info --format=json`
`nerdctl container list --format=json`
`nerdctl image list --format=json`

Was thinking that these could be generalized ?

It isn't ideal, sometimes the syntax is '{{json .}}'

```console
$ docker version --format json
json

```
